packages/pigeon/lib/dart_generator.dart

```diff
@@ -588,14 +592,62 @@ pigeonMap['${field.name}'] != null
   }
 }
 
-/// Generates Dart source code for test support libraries based on the
-/// given AST represented by [root], outputting the code to [sink].
+String? _findPubspecPath(String dartFilePath) {
+  try {
+    Directory dir = File(dartFilePath).parent;
+    String? pubspecPath;
+    while (pubspecPath == null) {
+      if (dir.existsSync()) {
+        final Iterable<String> pubspecPaths = dir
+            .listSync()
+            .map((FileSystemEntity e) => e.path)
+            .where((String path) => path.endsWith('pubspec.yaml'));
+        if (pubspecPaths.isNotEmpty) {
+          pubspecPath = pubspecPaths.first;
+        } else {
+          dir = dir.parent;
+        }
+      } else {
+        break;
+      }
+    }
+    return pubspecPath;
+  } catch (ex) {
+    return null;
+  }
+}
+
+String? _deducePackageName(String mainDartFile) {
+  final String? pubspecPath = _findPubspecPath(mainDartFile);
+  if (pubspecPath == null) {
+    return null;
+  }
+
+  final String text = File(pubspecPath).readAsStringSync();
+  final RegExp nameFinder = RegExp(r'^name:\s*(.*)');
+  final RegExpMatch? match = nameFinder.firstMatch(text);
+  if (match == null) {
+    return null;
+  }
+  return match.group(1);
+}
+
+String _posixify(String input) {
```

Comment:
This definitely needs a comment. `input` is very vague—in fact, nothing about this function signature indicates its about paths at all—and the fact that this would convert a relative path to an absolute path is non-obvious.

Reply:
done